_posts/2016/2016-04-05-tdd-with-java.md

```diff
@@ -0,0 +1,129 @@
+
+# Software Design
+
+The most important lesson for me in the TDD training was that the testability of the source code can be improved by the application of the following design principles:
+
+<table>
+<thead>
+<tr>
+<th>Principle</th>
+<th>Description</th>
+</tr>
+</thead>
+
+<tbody>
+<tr>
+<td>Loose coupling</td>
+<td>The dependencies between modules should be minimized.</td>
+</tr>
+
+<tr>
+<td>High cohesion</td>
+<td>Modules should be grouped by the data they are working with.</td>
+</tr>
+</tbody>
+</table>
+
+More information about this subject can be found under the keywords "[SOLID design principles](https://www.google.com/search?q%3Dsolid%2Bdesign%2Bprinciples)".
+
+## Design Pattern
+
+Loose coupling can be achieved by the application of the design pattern [Abstract Factory](http://www.tutorialspoint.com/design_pattern/abstract_factory_pattern.htm), [Bridge](http://www.tutorialspoint.com/design_pattern/bridge_pattern.htm), [Chain of Responsibility](http://www.tutorialspoint.com/design_pattern/chain_of_responsibility_pattern.htm), [Command](http://www.tutorialspoint.com/design_pattern/command_pattern.htm), [Facade](http://www.tutorialspoint.com/design_pattern/facade_pattern.htm), [Mediator](http://www.tutorialspoint.com/design_pattern/mediator_pattern.htm) and [Observer](http://www.tutorialspoint.com/design_pattern/observer_pattern.htm).
+
+## Refactoring
+
+High cohesion can be reached by the application of the refactoring pattern [Extract Class](http://refactoring.com/catalog/extractClass.html), [Extract Module](http://refactoring.com/catalog/extractModule.html), [Extract Subclass](http://refactoring.com/catalog/extractSubclass.html), [Extract Superclass](http://refactoring.com/catalog/extractSuperclass.html), [Pull Up Method](http://refactoring.com/catalog/pullUpMethod.html), [Push Down Method](http://refactoring.com/catalog/pushDownMethod.html), [Pull Up Field](http://refactoring.com/catalog/pullUpField.html), [Push Down Field](http://refactoring.com/catalog/pushDownField.html), [Move Field](http://refactoring.com/catalog/moveField.html) and [Move Method](http://refactoring.com/catalog/moveMethod.html).
+
+# Practical tips
+
+The advantage of training courses over books and videos is that they can include personalized, practical advices:
+-   Start working on a ticket by writing unit tests, always.
+-   Example is better than precept.
+-   The [Boy Scout Rule](http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule) can help to get started.
+-   At least 70% of the source code should be covered by unit tests, 80% is very good and more than 95% is perfect.
+-   TDD should be applied in a pragmatic way and not necessarily by the book.
+
+# Java Tools
+
+As TDD is all about writing unit tests, here is an overview the most commonly used Java tools for doing so:
+
+<table>
+<thead>
+<tr>
+<th>Framework</th>
+<th>Description</th>
+</tr>
+</thead>
+
+<tbody>
+<tr>
+<td>[JUnit](http://junit.org)</td>
```

Comment:
The links are not displayed correctly within a table in the preview from GitHub. I don't know how it should be correctly.